src/Frontend/Core.hs

```diff
@@ -113,13 +113,24 @@ class FormPageView p where
 class Page p where
     isPrivatePage :: p -> Bool
 
+    extraPageHeaders  :: p -> Html ()
+    extraPageHeaders _ = nil
+
 -- | Wrap anything that has 'ToHtml' and wrap it in an HTML body with complete page.
 data Frame body = Frame User body | PublicFrame body
   deriving (Functor)
 
--- FIXME: Replace it with enum type that represents the extra headers.
-data Frame' body = Frame' User (Html ()) body | PublicFrame' (Html ()) body
-  deriving (Functor)
+frameAlgebra :: (body -> a) -> Frame body -> a
+frameAlgebra f = \case (Frame _ b)     -> f b
+                       (PublicFrame b) -> f b
```

Comment:
Oh my! This could use the getter of frameBody lens if we were to derive it. See my next commit.